src/Coalesce.Framework.Persister.Elasticsearch/src/main/java/com/incadencecorp/coalesce/framework/persistance/elasticsearch/ElasticSearchPersistor.java

```diff
@@ -105,6 +112,55 @@ public boolean saveEntity(boolean allowRemoval, CoalesceEntity... entities) thro
         return true;
     }
 
+    private void executeCleanup(AbstractClient client, BulkRequest request, CoalesceEntity... entities)
+    {
+        Set<String> indices = new HashSet<>();
+        Set<String> recordKeys = new HashSet<>();
+
+        // Create List of Indices
+        for (DocWriteRequest task : request.requests())
+        {
+            recordKeys.add(task.id());
+            indices.add(task.index());
+        }
+
+        // Don't cleanup coalesce index
+        indices.remove(COALESCE_ENTITY_INDEX);
+
+        if (indices.size() > 0)
+        {
+            StopWatch watch = new StopWatch();
+            watch.start();
+
+            BoolQueryBuilder builder = QueryBuilders.boolQuery();
+
+            // Only records belonging to one of the entities being updated
+            for (CoalesceEntity entity : entities)
+            {
+                builder.should().add(QueryBuilders.matchQuery(ElasticSearchPersistor.ENTITY_KEY_COLUMN_NAME,
+                                                              entity.getKey()));
+            }
+
+            // Only records that are not being updated by this request
+            for (String key : recordKeys)
+            {
+                builder.mustNot().add(QueryBuilders.matchQuery("_id", key));
+            }
+
+            BulkByScrollResponse response = DeleteByQueryAction.INSTANCE.newRequestBuilder(client).filter(builder).
+                    source(indices.toArray(new String[indices.size()])).get();
```

Comment:
Any chance there could be more than 10K records returned from the query?   There are places in elastic search that have a hard limit at 10K and you have to do special paging to get more than that back.

Reply:
Ideally this would always return 0. If by chance someone update an entity 10K times without doing a merge you could technically exceed the 10K limit.

Reply:
Well ok so I created a test with 20K records and it does fail: 

NotSerializableExceptionWrapper[too_many_clauses: maxClauseCount is set to 1024]

Looks like its limited to 1024 clauses. Not sure how else to implement this behavior. We could of course catch this error and ignore it which means it would not prevent a user from creating or modifying the entity. However if you have an entity that exceeds 1024 records you would not have the protection of this logic.